text/0027-corejam.md

```diff
@@ -0,0 +1,706 @@
+```
+
+```rust
+type WorkOutputLen = ConstU32<1024>;
+type WorkOutput = BoundedVec<u8, WorkOutputLen>;
+fn refine(payload: WorkPayload) -> WorkOutput;
+```
+
+Both `refine` and `is_authorized` are only ever executed in-core. Within this environment, we need to ensure that we can interrupt computation not long after some well-specified limit and deterministically determine when an invocation of the VM exhausts this limit. Since the exact point at which interruption of computation need not be deterministic, it is expected to be executed by a streaming JIT transpiler with a means of approximate and overshooting interruption coupled with deterministic metering.
+
+Several host functions (largely in line with the host functions available to Parachain Validation Function code) are supplied. Two additional ones include:
+
+```rust
+/// Determine the preimage of `hash` utilizing the Relay-chain Storage pallet.
+fn lookup(hash: [u8; 32]) -> Vec<u8>;
+/// Determine the state root of the block at given `height`.
+fn state_root(height: u32) -> Option<[u8; 32]>;
+```
+
+Other host functions will allow for the possibility of executing a WebAssembly payload (for example, a Parachain Validation Function) or instantiating and entering a subordinate RISCV VM (for example for Actor Progressions).
+
+When applying `refine` from the client code, we must allow for the possibility that the VM exits unexpectedly or does not end. Validators are always rewarded for computing properly authorized Work Packages, including those which include such broken Work Items. But they must be able to report their broken state into the Relay-chain in order to collect their reward. Thus we define a type `WorkResult`:
+
+```rust
+enum WorkError {
+    Timeout,
+    Panic,
+}
+struct WorkResult {
+    class: WorkClass,
+    item_hash: [u8; 32],
+    result: Result<WorkOutput, WorkError>,
+    weight: Weight,
+}
+fn apply_refine(item: WorkItem) -> WorkResult;
+```
+
+The amount of weight used in executing the `refine` function is noted in the `WorkResult` value, and this is used later in order to help apportion on-chain weight (for the Join-Accumulate process) to the Work Classes whose items appear in the Work Packages.
+
+```rust
+struct WorkReport {
+    /// The hash of the underlying WorkPackage.
+    hash: WorkPackageHash,
+    /// The context of the underlying WorkPackage.
+    context: Context,
+    /// The core index of the attesting RCVG.
+    core_index: CoreIndex,
+    /// The results of the evaluation of the Items in the underlying Work Package.
+    results: BoundedVec<WorkResult, MaxWorkItemsInPackage>,
+}
+struct Attestation {
+    report: WorkReport,
+    validator: AccountId,
+    signature: Signature,
+}
+/// Since all RCVG members should be attesting to the same few Work Reports, it may
+/// make sense to send Attestations without the full underlying WorkReport, but only
+/// its hash.
+struct BareAttestation {
+    report_hash: WorkReportHash,
+    validator: AccountId,
```

Comment:
We currently have some more space-efficient packings of attestations which don't include the validator ID. It looks like this:

```rust
struct AttestedWorkReport {
  report: work_report,
  attestations: Vec<Signature>,
  group_attestations: BitVec, // indicates which validators from the group have a signature
}
```